include/ringbuffer_utils.hpp

```diff
@@ -360,13 +397,43 @@ inline const perf_event_header *perf_rb_read_event(RingBuffer &rb) {
   }
 
   uint64_t const tail_linear = tail & rb.mask;
-  auto *hdr = reinterpret_cast<perf_event_header *>(rb.data + tail_linear);
-  size_t const sz_sample = hdr->size;
+  auto *data_ptr = rb.data + tail_linear;
+
+  perf_event_header hdr_storage{};
+  perf_event_header *hdr_ptr = nullptr;
+  if (!rb.mirrored_mapping &&
+      tail_linear + sizeof(perf_event_header) > rb.data_size) {
+    size_t const first_chunk = rb.data_size - tail_linear;
+    memcpy(&hdr_storage, data_ptr, first_chunk);
+    memcpy(reinterpret_cast<std::byte *>(&hdr_storage) + first_chunk, rb.data,
+           sizeof(perf_event_header) - first_chunk);
+    hdr_ptr = &hdr_storage;
+  } else {
+    hdr_ptr = reinterpret_cast<perf_event_header *>(data_ptr);
+  }
+
+  size_t const sz_sample = hdr_ptr->size;
```

Comment:
Looks like you want to handle cases where `hdr_ptr->size < sizeof(perf_event_header)`
Then size to read should be `max(hdr_ptr->size, sizeof(perf_event_header)`, and that would allow removing `std::min` below + check against `first_chunk`